Add ConfigFlow for SPC

imduffy15 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->

Not yet but YAML support for SPC should be dropped in the future.

## Proposed change

 - Add config flow for SPC with automatic migration from YAML configuration
 - Add unique_ids to the sensors and alarm panels created by SPC
 - Add _attr_code_arm_required = false to match reality and allow arming from alexa
 - Upgrade to pyspcwebgw 0.6 release notes at https://github.com/pyspcwebgw/pyspcwebgw/releases

<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [x] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [?] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

imduffy15 commented:
The test `test_update_alarm_device` fails when the introduction of configflow, I'm unsure how to fix it and would appreciate input.

Thanks.

imduffy15 commented:
https://github.com/pyspcwebgw/pyspcwebgw/releases shows the changes for pyspcwebgw, nothing major, bumped it to get the version information to make the device entry look nice.

gjohansson-ST reviewed (changes requested):
It seems you are missing to create an issue on the migration of yaml and also `async_setup_entry` seems missing in atleast `alarm_control_panel` so it would never load.

imduffy15 commented:
> It seems you are missing to create an issue on the migration of yaml and also `async_setup_entry` seems missing in atleast `alarm_control_panel` so it would never load.

Hey! Thanks for taking a look really appreciate it, new to the code base. 

Can create an issue that is no problem. 

I'm confused by the second bit, isn't the following responsible for creating the alarm panel? It appeared when running locally. 

'''

    # create a separate alarm panel for each area
    hass.async_create_task(
        discovery.async_load_platform(
            hass, Platform.ALARM_CONTROL_PANEL, DOMAIN, {}, hass.data[DOMAIN]
        )
    )
'''

I'm very confused as to why the pytest fails. If I am following correctly it uses the yaml setup and waits for any async tasks to create. The async task to run the aync_setup_entity should run and create the alarm panel and sensors but this doesn't seem to occur in the tests.

gjohansson-ST commented:
> I'm confused by the second bit, isn't the following responsible for creating the alarm panel? It appeared when running locally.

Wasn't looking careful enough apparently.
With the implementation of a config flow to make it UI based you should import any YAML to a config entry and use it to load the integration from their on.
This PR would not be approved based on having both UI and YAML setup.

imduffy15 commented:
Hi @gjohansson-ST unless im misunderstanding something the code is already in place to do the import:

async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
    """Set up the SPC component."""
    if DOMAIN in config:
        hass.async_create_task(
            hass.config_entries.flow.async_init(
                DOMAIN, context={"source": SOURCE_IMPORT}, data=config[DOMAIN]
            )
        )

    return True

Starts the configflow import process. 

Tested this locally by creating yaml and seeing if a config entity was created... it was.

gjohansson-ST commented on homeassistant/components/spc/__init__.py:
```diff
@@ -62,15 +77,31 @@ async def async_upate_callback(spc_object):
     # add sensor devices for each zone (typically motion/fire/door sensors)
     hass.async_create_task(
-        discovery.async_load_platform(hass, Platform.BINARY_SENSOR, DOMAIN, {}, config)
+        discovery.async_load_platform(
```
You should use `hass.config_entries.async_forward_entry_setups` instead to load platforms.

imduffy15 replied:
Thank you for the pointers!

gjohansson-ST reviewed (changes requested):
I now took a proper look.
So the import should be fine but the way of loading platforms is not. I suggest you look in the code-base with my snip as reference point to search for

imduffy15 commented:
![spc](https://user-images.githubusercontent.com/1243435/218881394-301784cf-9a8b-4dae-b502-d69fbe5dfa37.gif)

gjohansson-ST commented on homeassistant/components/spc/__init__.py:
```diff
@@ -40,20 +42,35 @@
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up SPC from a config entry."""
+
+    hass.data.setdefault(DOMAIN, {})
```
Is this used somewhere?

gjohansson-ST commented on homeassistant/components/spc/__init__.py:
```diff
@@ -62,19 +79,27 @@ async def async_upate_callback(spc_object):
-    hass.async_create_task(
-        discovery.async_load_platform(hass, Platform.BINARY_SENSOR, DOMAIN, {}, config)
+    device_registry = dr.async_get(hass)
+    device_registry.async_get_or_create(
```
Should this perhaps not be created as a device from the `alarm_control_panel`?

imduffy15 replied:
I'm not sure, I haven't seen that done in other places, it all seems to be part of __init__ or a coordinator:

```
$ rg "async_get_or_create" | sort | uniq | sed -e '/test/d'
agent_dvr/__init__.py:    device_registry.async_get_or_create(
apple_tv/__init__.py:        device_registry.async_get_or_create(
auth/__init__.py:        user = await hass.auth.async_get_or_create_user(credential)
axis/device.py:        device_registry.async_get_or_create(
bluetooth/__init__.py:    dr.async_get(hass).async_get_or_create(
bond/__init__.py:    device_registry.async_get_or_create(
bosch_shc/__init__.py:    device_registry.async_get_or_create(
bthome/config_flow.py:                    return self._async_get_or_create_entry(bindkey)
bthome/config_flow.py:            return self._async_get_or_create_entry()
bthome/config_flow.py:    def _async_get_or_create_entry(self, bindkey=None):
config/auth_provider_homeassistant.py:    credentials = await provider.async_get_or_create_credentials(
control4/__init__.py:    device_registry.async_get_or_create(
deconz/deconz_event.py:        entry = device_registry.async_get_or_create(
deconz/gateway.py:        device_registry.async_get_or_create(
dlna_dmr/media_player.py:        device_entry = dev_reg.async_get_or_create(
dlna_dmr/media_player.py:        ent_reg.async_get_or_create(
eight_sleep/__init__.py:        dev_reg.async_get_or_create(
eight_sleep/__init__.py:    dev_reg.async_get_or_create(
esphome/__init__.py:    device_entry = device_registry.async_get_or_create(
fibaro/__init__.py:    device_registry.async_get_or_create(
firmata/__init__.py:    device_registry.async_get_or_create(
fronius/__init__.py:        device_registry.async_get_or_create(
google_assistant/__init__.py:    device_registry.async_get_or_create(
guardian/__init__.py:        device = dev_reg.async_get_or_create(
hassio/__init__.py:        dev_reg.async_get_or_create(config_entry_id=entry_id, **params)
hassio/__init__.py:    dev_reg.async_get_or_create(config_entry_id=entry_id, **params)
home_connect/__init__.py:            device_entry = device_registry.async_get_or_create(
homekit/__init__.py:        dev_reg.async_get_or_create(
homekit_controller/connection.py:            device = device_registry.async_get_or_create(
homekit_controller/device_trigger.py:        trigger = async_get_or_create_trigger_source(conn.hass, device_id)
homekit_controller/device_trigger.py:    return async_get_or_create_trigger_source(hass, device_id).async_attach_trigger(
homekit_controller/device_trigger.py:def async_get_or_create_trigger_source(
homematicip_cloud/__init__.py:    device_registry.async_get_or_create(
huawei_lte/__init__.py:        device_registry.async_get_or_create(
hue/__init__.py:        device_registry.async_get_or_create(
hue/v1/hue_event.py:        entry = device_registry.async_get_or_create(
hue/v2/device.py:        return dev_reg.async_get_or_create(config_entry_id=entry.entry_id, **params)
insteon/__init__.py:    device_registry.async_get_or_create(
isy994/__init__.py:    _async_get_or_create_isy_device_in_registry(hass, entry, isy)
isy994/__init__.py:    device_registry.async_get_or_create(
isy994/__init__.py:def _async_get_or_create_isy_device_in_registry(
konnected/panel.py:        device_registry.async_get_or_create(
lcn/helpers.py:        device_registry.async_get_or_create(
lcn/helpers.py:    device_registry.async_get_or_create(
livisi/__init__.py:    device_registry.async_get_or_create(
lutron_caseta/__init__.py:            dr_device = device_registry.async_get_or_create(
lutron_caseta/__init__.py:    device_registry.async_get_or_create(**device_args, config_entry_id=config_entry_id)
matter/adapter.py:        dr.async_get(self.hass).async_get_or_create(
mikrotik/__init__.py:    device_registry.async_get_or_create(
mobile_app/__init__.py:    device = device_registry.async_get_or_create(
mobile_app/config_flow.py:        devt_entry = entity_registry.async_get_or_create(
mobile_app/webhook.py:    device_registry.async_get_or_create(
motion_blinds/__init__.py:        device_registry.async_get_or_create(
motioneye/__init__.py:    device = device_registry.async_get_or_create(
mqtt/mixins.py:            device_registry.async_get_or_create(
mqtt/mixins.py:        device = device_registry.async_get_or_create(**update_device_info)
nanoleaf/__init__.py:        device_entry = device_registry.async_get_or_create(
netgear/__init__.py:    device_registry.async_get_or_create(
nibe_heatpump/__init__.py:    device_entry = reg.async_get_or_create(
nightscout/__init__.py:    device_registry.async_get_or_create(
nobo_hub/__init__.py:    dev_reg.async_get_or_create(
notion/__init__.py:    device_registry.async_get_or_create(
nuki/__init__.py:    dev_reg.async_get_or_create(
nut/__init__.py:    device_registry.async_get_or_create(
onboarding/views.py:            credentials = await provider.async_get_or_create_credentials(
onewire/onewirehub.py:            device_registry.async_get_or_create(
opentherm_gw/__init__.py:        gw_dev = dev_reg.async_get_or_create(
overkiz/__init__.py:        device_registry.async_get_or_create(
plugwise/gateway.py:    device_registry.async_get_or_create(
ps4/__init__.py:                registry.async_get_or_create(
rachio/__init__.py:        webhook_url = await async_get_or_create_registered_webhook_id_and_url(
rachio/__init__.py:    async_get_or_create_registered_webhook_id_and_url,
rachio/webhooks.py:async def async_get_or_create_registered_webhook_id_and_url(
renault/renault_hub.py:        device_registry.async_get_or_create(
roon/__init__.py:    device_registry.async_get_or_create(
ruckus_unleashed/__init__.py:        registry.async_get_or_create(
sfr_box/__init__.py:    device_registry.async_get_or_create(
shelly/coordinator.py:        device_entry = dev_reg.async_get_or_create(
sia/hub.py:            device_registry.async_get_or_create(
simplisafe/__init__.py:    base_station = device_registry.async_get_or_create(
smarttub/controller.py:            device_registry.async_get_or_create(
sonos/switch.py:        new_device = device_registry.async_get_or_create(
spc/__init__.py:    device_registry.async_get_or_create(
starline/__init__.py:        device_registry.async_get_or_create(
switcher_kis/__init__.py:        dev_reg.async_get_or_create(
syncthru/__init__.py:    device_registry.async_get_or_create(
tasmota/__init__.py:    device_registry.async_get_or_create(
tellduslive/__init__.py:        dev_reg.async_get_or_create(
toon/__init__.py:    device_registry.async_get_or_create(
tradfri/__init__.py:    dev_reg.async_get_or_create(
tuya/__init__.py:        device_registry.async_get_or_create(
unifi/controller.py:        device_registry.async_get_or_create(
upnp/__init__.py:        device_entry = dev_registry.async_get_or_create(
utility_meter/__init__.py:        entity_entry = entity_registry.async_get_or_create(
wemo/wemo_device.py:    entry = device_registry.async_get_or_create(
xiaomi_aqara/__init__.py:    device_registry.async_get_or_create(
xiaomi_ble/__init__.py:            device = device_registry.async_get_or_create(
xiaomi_ble/config_flow.py:                    return self._async_get_or_create_entry(bindkey)
xiaomi_ble/config_flow.py:            return self._async_get_or_create_entry()
xiaomi_ble/config_flow.py:    def _async_get_or_create_entry(self, bindkey=None):
xiaomi_miio/__init__.py:    device_registry.async_get_or_create(
zha/__init__.py:    device_registry.async_get_or_create(
zha/core/gateway.py:            device_registry_device = self.ha_device_registry.async_get_or_create(
zha/core/gateway.py:            zha_device = self._async_get_or_create_device(zigpy_device, restored=True)
zha/core/gateway.py:            zha_group = self._async_get_or_create_group(group)
zha/core/gateway.py:        zha_device = self._async_get_or_create_device(device)
zha/core/gateway.py:        zha_group = self._async_get_or_create_group(zigpy_group)
zha/core/gateway.py:    def _async_get_or_create_device(
zha/core/gateway.py:    def _async_get_or_create_group(self, zigpy_group: zigpy.group.Group) -> ZHAGroup:
zwave_js/__init__.py:        device = self.dev_reg.async_get_or_create(
```

gjohansson-ST replied:
I guess it depends if you intend to have it as a device for a hub or if it's a device representing the alarm panel.
Probably the first one which should be ok then but wanted to ask.

imduffy15 replied:
OK. Resolving.

gjohansson-ST commented on homeassistant/components/spc/__init__.py:
```diff
@@ -62,19 +79,27 @@ async def async_upate_callback(spc_object):
         _LOGGER.error("Failed to load area/zone information from SPC")
```
This should raise [ConfigEntryNotReady](https://github.com/home-assistant/core/blob/dev/homeassistant/exceptions.py#L119)

gjohansson-ST commented on homeassistant/components/spc/alarm_control_panel.py:
```diff
@@ -95,6 +93,11 @@ def state(self) -> str | None:
         return _get_alarm_state(self._area)
 
+    @property
+    def unique_id(self) -> str:
```
You can set this as `_attr_unique_id` already on __init__

gjohansson-ST commented on homeassistant/components/spc/binary_sensor.py:
```diff
@@ -83,3 +80,8 @@ def is_on(self) -> bool:
         return _get_device_class(self._zone.type)
+
+    @property
+    def unique_id(self) -> str:
```
Use `_attr_unique_id`

gjohansson-ST commented on homeassistant/components/spc/config_flow.py:
```diff
@@ -0,0 +1,100 @@
+class SPCConfigFlow(ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for SPC."""
+
+    VERSION = 2
```
```suggestion
```

gjohansson-ST commented on homeassistant/components/spc/config_flow.py:
```diff
@@ -0,0 +1,100 @@
+        """Import a config entry from configuration.yaml."""
+        for entry in self._async_current_entries():
+            if entry.data[CONF_API_URL] == import_config[CONF_API_URL]:
+                _LOGGER.warning(
+                    "Already configured. This yaml configuration has already been imported. Please remove it"
+                )
```
```suggestion
```
This logging doesn't add any value

gjohansson-ST commented on homeassistant/components/spc/config_flow.py:
```diff
@@ -0,0 +1,100 @@
+            )
+            try:
+                spc = await _validate_input(self.hass, user_input)
+                return self.async_create_entry(
+                    title=f"{spc.info['type']} - {spc.info['sn']}",
+                    data=user_input,
+                )
```
Move this outside the try-block

imduffy15 replied:
That wouldn't make sense I'm afraid, errors are populated in the below exemption and then it falls through to the return to show the form. If this is brought outside of the try-block then this code would be executed with no spc populated.

gjohansson-ST replied:
Move it to else (try-except-else)

imduffy15 replied:
I never knew you could do an else on those, thank you for teaching me new things.

gjohansson-ST commented on homeassistant/components/spc/config_flow.py:
```diff
@@ -0,0 +1,100 @@
+            except SpcConnectionFailure:
+                errors = {"base": "failed_to_connect"}
+            except Exception:  # pylint: disable=broad-except
+                _LOGGER.exception("Unexpected exception")
```
```suggestion
```

gjohansson-ST commented on homeassistant/components/spc/config_flow.py:
```diff
@@ -0,0 +1,100 @@
+                _LOGGER.exception("Unexpected exception")
+                return self.async_abort(reason="unknown")
+
+        data_schema = self._get_user_data_schema()
```
I think you should define this as a constant in the top.

gjohansson-ST commented on homeassistant/components/spc/const.py:
```diff
@@ -0,0 +1,21 @@
+SIGNAL_UPDATE_ALARM = "spc_update_alarm_{}"
+SIGNAL_UPDATE_SENSOR = "spc_update_sensor_{}"
+
+CONFIG_SCHEMA = {
```
Does not look like it's used and should anyhow be where it's used and not here

gjohansson-ST replied:
Move this to where it's used e.g. config_flow

gjohansson-ST reviewed (changes requested):
Some steps to get going

imduffy15 commented on homeassistant/components/spc/__init__.py:
```diff
@@ -62,19 +79,27 @@ async def async_upate_callback(spc_object):
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    if unload_ok := await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
+        hass.data[DOMAIN].pop(entry.entry_id)
```
Fixed! Thank you, had it in from the skeleton.

gjohansson-ST commented on homeassistant/components/spc/__init__.py:
```diff
@@ -40,41 +43,58 @@
+        async_callback=lambda spc_object: _async_upate_callback(hass, spc_object),
     )
 
     hass.data[DATA_API] = spc
```
Looks like this would be a problem if you setup multiple entries of spc

imduffy15 replied:
The `spc_object.id` would be different so should be OK.

imduffy15 replied:
Changed it to make use of the entry.entity_id

gjohansson-ST commented on homeassistant/components/spc/config_flow.py:
```diff
@@ -0,0 +1,87 @@
+
+        if user_input is not None:
+            self._async_abort_entries_match(
+                {
+                    CONF_API_URL: user_input[CONF_API_URL],
+                    CONF_WS_URL: user_input[CONF_WS_URL],
+                }
```
```suggestion
                user_input
```

gjohansson-ST commented on homeassistant/components/spc/config_flow.py:
```diff
@@ -0,0 +1,87 @@
+                spc = await _validate_input(self.hass, user_input)
+            except SpcConnectionFailure:
+                errors = {"base": "failed_to_connect"}
+            except Exception:  # pylint: disable=broad-except
+                return self.async_abort(reason="unknown")
```
Nothing here that would raise anything else than `SpcConnectionFailure`?

imduffy15 replied:
SpcConnectionFailure caught on 73.

This is a catch all unknown, copied it from another component.

gjohansson-ST commented on homeassistant/components/spc/manifest.json:
```diff
@@ -2,8 +2,9 @@
   "iot_class": "local_push",
   "loggers": ["pyspcwebgw"],
-  "requirements": ["pyspcwebgw==0.4.0"]
+  "requirements": ["pyspcwebgw==0.6.0"]
```
There should be a link in the PR description to the release or something for this dependency upgrade

imduffy15 replied:
covered at https://github.com/home-assistant/core/pull/87915#issuecomment-1426903807

gjohansson-ST reviewed (changes requested):
Few more comments

imduffy15 committed:
SPC enhancements

 - Add config flow for SPC with automatic migration from YAML configuration
 - Add unique_ids to the sensors and alarm panels created by SPC
 - Add _attr_code_arm_required = false to match reality and allow arming from alexa
 - Upgrade to pyspcwebgw 0.6 release notes at https://github.com/pyspcwebgw/pyspcwebgw/releases

epenet commented:
Notes:
> Add config flow for SPC with automatic migration from YAML configuration
> Add unique_ids to the sensors and alarm panels created by SPC
> Add _attr_code_arm_required = false to match reality and allow arming from alexa
> Upgrade to pyspcwebgw 0.6 release notes at https://github.com/pyspcwebgw/pyspcwebgw/releases

Should these not be four separate PR?
Config flow tends to be a big PR with quite a lot of back-and-forth, but the other three could probably be 3 preliminary or follow-up PRs.

imduffy15 commented:
> Notes:
> 
> > Add config flow for SPC with automatic migration from YAML configuration
> 
> > Add unique_ids to the sensors and alarm panels created by SPC
> 
> > Add _attr_code_arm_required = false to match reality and allow arming from alexa
> 
> > Upgrade to pyspcwebgw 0.6 release notes at https://github.com/pyspcwebgw/pyspcwebgw/releases
> 
> 
> 
> Should these not be four separate PR?
> 
> Config flow tends to be a big PR with quite a lot of back-and-forth, but the other three could probably be 3 preliminary or follow-up PRs.

Happy to just abandon this. Contributing is becoming too difficult.

epenet commented:
> > Notes:
> > > Add config flow for SPC with automatic migration from YAML configuration
> > 
> > 
> > > Add unique_ids to the sensors and alarm panels created by SPC
> > 
> > 
> > > Add _attr_code_arm_required = false to match reality and allow arming from alexa
> > 
> > 
> > > Upgrade to pyspcwebgw 0.6 release notes at https://github.com/pyspcwebgw/pyspcwebgw/releases
> > 
> > 
> > Should these not be four separate PR?
> > Config flow tends to be a big PR with quite a lot of back-and-forth, but the other three could probably be 3 preliminary or follow-up PRs.
> 
> Happy to just abandon this. Contributing is becoming too difficult.

`config-flow` is quite often the most difficult - that's why I'm suggesting that you move the other code changes to small individual PRs that can be reviewed and merged independently.

imduffy15 commented:
> > > Notes:
> 
> > > > Add config flow for SPC with automatic migration from YAML configuration
> 
> > > 
> 
> > > 
> 
> > > > Add unique_ids to the sensors and alarm panels created by SPC
> 
> > > 
> 
> > > 
> 
> > > > Add _attr_code_arm_required = false to match reality and allow arming from alexa
> 
> > > 
> 
> > > 
> 
> > > > Upgrade to pyspcwebgw 0.6 release notes at https://github.com/pyspcwebgw/pyspcwebgw/releases
> 
> > > 
> 
> > > 
> 
> > > Should these not be four separate PR?
> 
> > > Config flow tends to be a big PR with quite a lot of back-and-forth, but the other three could probably be 3 preliminary or follow-up PRs.
> 
> > 
> 
> > Happy to just abandon this. Contributing is becoming too difficult.
> 
> 
> 
> `config-flow` is quite often the most difficult - that's why I'm suggesting that you move the other code changes to small individual PRs that can be reviewed and merged independently.

We can accept as is or I can close. 

Contributing to home assistant has become far too difficult - and I say that as someone who is a PMC for an Apache foundation project. At this point it is easier to maintain my own fork and rebase on each release. 

While I understand the ask here is small we are 15 days into this review with ~40 comments having being shared. 

Additionally, when attempting to make small changes to components with configflow in the past they have been rejected. https://github.com/home-assistant/home-assistant.io/pull/15793

Test coverage is there. Review feedback has been implemented. Cla has been signed. All ci is passing. Recording demonstrating the functionality has been provided. This can be accepted as is or I can close the pr. 

Let me know how you wish to proceed.

epenet commented:
See https://developers.home-assistant.io/docs/review-process#creating-the-perfect-pr
> Make your PRs as small as possible. A PR should only refactor one thing, fix one thing, add one feature, or adjust a single subject in the documentation. If you want to change multiple things, please create multiple PRs. Smaller PRs have a smaller scope, need less time to review, conflict less often, and generally need fewer review iterations.

>Only change one thing at a time. This is the same as the previous point but a bit more specific. It is tempting to improve those one or two lines you've noticed nearby, but please don't. Put those in a separate PR. Unrelated changes in your PR are distracting and often lead to questions. In contrast, in an independent PR, it would be a quick and simple review and merge.

Anyway I will not interfere more with the review from @gjohansson-ST

imduffy15 committed:
Merge branch 'dev' into spc-configflow

imduffy15 committed:
Merge branch 'dev' into spc-configflow

imduffy15 commented:
Closing this as its gone stale and now fails code coverage. Not putting any more time into it.

Thanks for destroy the PR @epenet :/
